mesonbuild/backend/ninjabackend.py

```diff
@@ -1131,9 +1131,14 @@ def generate_vala_compile(self, target, outfile):
             # file is outside the build directory, the path components will be
             # stripped and just the basename will be used.
             if isinstance(gensrc, (build.CustomTarget, build.GeneratedList)) or gensrc.is_built:
-                vala_c_file = os.path.splitext(vala_file)[0] + '.c'
-            else:
                 vala_c_file = os.path.splitext(os.path.basename(vala_file))[0] + '.c'
+            else:
+                realpath = os.path.abspath(os.path.join(self.environment.get_build_dir(), vala_file))
+                if (realpath.startswith(os.path.join(self.environment.get_source_dir(), target.get_subdir()))):
+                    relpath = os.path.relpath(realpath, os.path.join(self.environment.get_source_dir(), target.get_subdir()))
+                    vala_c_file = os.path.join(os.path.dirname(relpath), os.path.splitext(os.path.basename(vala_file))[0] + '.c')
+                else:
+                    vala_c_file = os.path.splitext(os.path.basename(vala_file))[0] + '.c'
```

Comment:
This code is unnecessarily complicated. You should just do something like:

```py
vala_c_file = vala_file.replace('/', '_').replace('\\', '_') + '.c'
```

We do similar things elsewhere.

Reply:
That only works on the Meson side, similar to .o files. The problem, however, is the Vala compiler does not have any switches for amending the names of the output C files. `--output` is for binary files produced using the `valac` command. So using a `replace` like would cause a mismatch between what `valac` produces and what Meson is expecting.

Reply:
Do you think you can file a bug about that with Vala? They really should have a flag like that or out-of-source builds become difficult like this.